src/core_read_parameter_file.c

```diff
@@ -23,196 +33,224 @@ int read_parameter_file(const int ThisTask, const char *fname, struct params *ru
     ParamID[NParam++] = DOUBLE;
 
-    strcpy(ParamTag[NParam], "EtaSN");
+    strncpy(ParamTag[NParam], "EtaSN", MAXTAGLEN);
     ParamAddr[NParam] = &(run_params->EtaSN);
     ParamID[NParam++] = DOUBLE;
 
-    strcpy(ParamTag[NParam], "Yield");
+    strncpy(ParamTag[NParam], "Yield", MAXTAGLEN);
     ParamAddr[NParam] = &(run_params->Yield);
     ParamID[NParam++] = DOUBLE;
 
-    strcpy(ParamTag[NParam], "FracZleaveDisk");
+    strncpy(ParamTag[NParam], "FracZleaveDisk", MAXTAGLEN);
     ParamAddr[NParam] = &(run_params->FracZleaveDisk);
     ParamID[NParam++] = DOUBLE;
 
-    strcpy(ParamTag[NParam], "SfrEfficiency");
+    strncpy(ParamTag[NParam], "SfrEfficiency", MAXTAGLEN);
     ParamAddr[NParam] = &(run_params->SfrEfficiency);
     ParamID[NParam++] = DOUBLE;
 
-    strcpy(ParamTag[NParam], "FeedbackReheatingEpsilon");
+    strncpy(ParamTag[NParam], "FeedbackReheatingEpsilon", MAXTAGLEN);
     ParamAddr[NParam] = &(run_params->FeedbackReheatingEpsilon);
     ParamID[NParam++] = DOUBLE;
 
-    strcpy(ParamTag[NParam], "FeedbackEjectionEfficiency");
+    strncpy(ParamTag[NParam], "FeedbackEjectionEfficiency", MAXTAGLEN);
     ParamAddr[NParam] = &(run_params->FeedbackEjectionEfficiency);
     ParamID[NParam++] = DOUBLE;
 
-    strcpy(ParamTag[NParam], "BlackHoleGrowthRate");
+    strncpy(ParamTag[NParam], "BlackHoleGrowthRate", MAXTAGLEN);
     ParamAddr[NParam] = &(run_params->BlackHoleGrowthRate);
     ParamID[NParam++] = DOUBLE;
 
-    strcpy(ParamTag[NParam], "RadioModeEfficiency");
+    strncpy(ParamTag[NParam], "RadioModeEfficiency", MAXTAGLEN);
     ParamAddr[NParam] = &(run_params->RadioModeEfficiency);
     ParamID[NParam++] = DOUBLE;
 
-    strcpy(ParamTag[NParam], "QuasarModeEfficiency");
+    strncpy(ParamTag[NParam], "QuasarModeEfficiency", MAXTAGLEN);
     ParamAddr[NParam] = &(run_params->QuasarModeEfficiency);
     ParamID[NParam++] = DOUBLE;
 
-    strcpy(ParamTag[NParam], "Reionization_z0");
+    strncpy(ParamTag[NParam], "Reionization_z0", MAXTAGLEN);
     ParamAddr[NParam] = &(run_params->Reionization_z0);
     ParamID[NParam++] = DOUBLE;
 
-    strcpy(ParamTag[NParam], "Reionization_zr");
+    strncpy(ParamTag[NParam], "Reionization_zr", MAXTAGLEN);
     ParamAddr[NParam] = &(run_params->Reionization_zr);
     ParamID[NParam++] = DOUBLE;
 
-    strcpy(ParamTag[NParam], "ThresholdSatDisruption");
+    strncpy(ParamTag[NParam], "ThresholdSatDisruption", MAXTAGLEN);
     ParamAddr[NParam] = &(run_params->ThresholdSatDisruption);
     ParamID[NParam++] = DOUBLE;
 
-    strcpy(ParamTag[NParam], "NumOutputs");
+    strncpy(ParamTag[NParam], "NumOutputs", MAXTAGLEN);
     ParamAddr[NParam] = &(run_params->NOUT);
     ParamID[NParam++] = INT;
 
-    strcpy(ParamTag[NParam], "OutputFormat");
+    strncpy(ParamTag[NParam], "OutputFormat", MAXTAGLEN);
     ParamAddr[NParam] = my_outputformat;
     ParamID[NParam++] = STRING;
 
+    strncpy(ParamTag[NParam], "ForestDistributionScheme", MAXTAGLEN);
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Easily used incorrectly; doesn't always \0-terminate or check for invalid pointers [MS-banned] (CWE-120).  Risk is low because the source is a constant string.](https://app.codacy.com/manual/jacobseiler/sage-model/pullRequest?prid=4620074)